MCQ and MSQ question: fix cancel question feature #7934

tshradheya opened the pull request:
Fixes #7934 

**Outline of Solution**

- Added `if-else` conditon to check for cases when `generateMcqOptionsCheckbox`  and `generateMsqOptionsCheckbox `checkbox is clicked or not. The initial `generateOptionsChecbox` was the wrong id.
- Added tests which fail with previous logic and pass with current logic

tshradheya commented:
I am thinking that the following lines could be removed as the lines added are kind of duplicate:

https://github.com/tshradheya/teammates/blob/b757563de6f82c1a273c1aef1176847548d57b3b/src/main/webapp/dev/js/main/instructorFeedbackEdit.js#L536-L546

But I am unsure if it will cause any regression. Any senior dev can confirm?

wkurniawan07 commented:
@tshradheya you can try it yourself and see how it goes.

tshradheya commented:
@wkurniawan07 Removed the duplicate part. Since the tests are still passing and on doing some manual testing it still works as expected.

As for the PR, should I be adding tests as well?

wkurniawan07 commented:
@tshradheya you should; in fact, you should add some that will fail before you implemented the fix, otherwise how can you be sure than you are actually fixing something?

tshradheya commented:
@wkurniawan07 As suggested, in the following commit I reverted to old code and added tests which led to it failing on my local machine. https://github.com/TEAMMATES/teammates/pull/8472/commits/a2d94500c502780cb385cfd80232c6d9ac43ad71 

Then I added the code which fixed the bug and it passed on my local machine as well. https://github.com/TEAMMATES/teammates/pull/8472/commits/433e0d23fcccb63df67b6c3f7f620b01b1540f37

However, I noticed that even though https://github.com/TEAMMATES/teammates/pull/8472/commits/a2d94500c502780cb385cfd80232c6d9ac43ad71 failed on my local machine, it passed on Travis. 

It fails correctly on my device at line 74: `        assertFalse(feedbackEditPage.isElementVisible("mcqChoiceTable--1"));
`
![image](https://user-images.githubusercontent.com/12574756/36383549-66c4ac5c-15c7-11e8-9704-7b5f72f51881.png)

tshradheya commented:
Ready for Review

Crphang commented on src/main/webapp/dev/js/main/instructorFeedbackEdit.js:
```diff
@@ -535,15 +535,19 @@ function enableNewQuestion() {
 
     moveAssignWeightsCheckbox($newQuestionTable.find(`#rubricAssignWeights-${NEW_QUESTION}`));
 
-    if ($(`#generateOptionsCheckbox-${NEW_QUESTION}`).prop('checked')) {
```
Is generateOptionsCheckbox an id in any other type of questions?

Crphang replied:
Because amongst your changes it doesn't seem like you changed any jsp/html file. So it could either break something or it might be that the initial id was a typo.

tshradheya replied:
The initial id `generateOptionsCheckbox` was a typo which was the reason why adding question again after canceling didn't trigger the check. Doing a project-wide search, `generateOptionsCheckbox` is currently only used in 3 places and all those are part of `instructorHelpSession.jsp` ( I am assuming its outdated code)

Crphang replied:
@tshradheya LGTM

tshradheya commented:
Gentle reminder for review

Crphang commented:
@wkurniawan07 I've reviewed couple days ago. LGTM :+1:

wkurniawan07 commented:
@Crphang looks like you're not familiar with GitHub review feature yet; go to "Files changed" -> "Review changes" -> "Approve" if you approve of the changes.

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMcqQuestionUiTest.java:
```diff
@@ -59,6 +59,23 @@ public void testNewQuestionFrame() {
+        assertFalse(feedbackEditPage.isElementVisible("mcqChoiceTable--1"));
+        assertTrue(feedbackEditPage.isElementEnabled("mcqGenerateForSelect--1"));
+
+        feedbackEditPage.clickGenerateMcqOptionsCheckbox(-1); //reset back to normal condition
```
Do we have a constant (might be defined somewhere) for the magic `-1`? (I know it is the index of a new question, but it is not really friendly to humans). Besides, you can also try to improve the comment. `Make the checkbox unchecked` might be better than `reset back to normal condition`

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMsqQuestionUiTest.java:
```diff
@@ -59,6 +59,23 @@ public void testNewQuestionFrame() {
         feedbackEditPage.clickNewQuestionButton();
         feedbackEditPage.selectNewQuestionType("MSQ");
         assertTrue(feedbackEditPage.verifyNewMsqQuestionFormIsDisplayed());
+
```
Same here

xpdavid commented on src/main/webapp/dev/js/main/instructorFeedbackEdit.js:
```diff
@@ -537,15 +537,19 @@ function enableNewQuestion() {
     moveAssignWeightsCheckbox($newQuestionTable.find(`#rubricAssignWeights-${NEW_QUESTION}`));
 
-    if ($(`#generateOptionsCheckbox-${NEW_QUESTION}`).prop('checked')) {
+    if ($(`#generateMcqOptionsCheckbox-${NEW_QUESTION}`).prop('checked')) {
```
After you have added this, we will have three (almost) identical code in `disableQuestion` `enableQuestion` and `enableNewQuestion`. See if you can extract these lines into a common method.

tshradheya replied:
It's possible to extract from `enableQuestion` and `enableNewQuestion` only.

The other one although possible will result in extra code in the function which won't be used by other two calls. Also the logic isn't exactly the same, so thee will still be code duplication

xpdavid replied:
@tshradheya See `toggleMcqGeneratedOptions()` in `questionMcq.js`. You can make use of that method actually instead of writing it again.

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMsqQuestionUiTest.java:
```diff
@@ -59,6 +59,23 @@ public void testNewQuestionFrame() {
+        assertFalse(feedbackEditPage.isElementVisible("msqChoiceTable--1"));
+        assertTrue(feedbackEditPage.isElementEnabled("msqGenerateForSelect--1"));
+
+        feedbackEditPage.clickGenerateMsqOptionsCheckbox(-1); //reset back to normal condition
```
Same here. Magic `-1` :P

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMcqQuestionUiTest.java:
```diff
@@ -59,6 +59,23 @@ public void testNewQuestionFrame() {
         feedbackEditPage.selectNewQuestionType("MCQ");
         assertTrue(feedbackEditPage.verifyNewMcqQuestionFormIsDisplayed());
+
+        feedbackEditPage.clickGenerateMcqOptionsCheckbox(-1);
```
I believe the test case should start here?
`______TS("MSQ: Check UI after cancelling and add new MCQ question again");`

tshradheya commented:
Made necessary changes. Ready for Review

tshradheya commented:
Avoided code duplication as suggested. Tested if the bug is fixed with code changes and it works perfect. 
Ready for Review

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMcqQuestionUiTest.java:
```diff
@@ -59,6 +62,23 @@ public void testNewQuestionFrame() {
+        assertFalse(feedbackEditPage.isElementVisible("mcqChoiceTable--1"));
+        assertTrue(feedbackEditPage.isElementEnabled("mcqGenerateForSelect--1"));
+
+        feedbackEditPage.clickGenerateMcqOptionsCheckbox(-QUESTION_INDEX_ONE); //Make the generate options checkbox unchecked
```
Why not just define the new question index as `-1` like what we do in JS file `instructorFeedbackEdit.js`.
``` javascript
const NEW_QUESTION = -1;
```

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMsqQuestionUiTest.java:
```diff
@@ -162,10 +182,10 @@ public void testEditQuestionAction() throws Exception {
         feedbackEditPage.clickEditQuestionButton(1);
-        feedbackEditPage.fillQuestionTextBox("edited msq qn text", 1);
-        feedbackEditPage.fillQuestionDescription("more details", 1);
+        feedbackEditPage.fillQuestionTextBox("edited msq qn text", QUESTION_INDEX_ONE);
```
There is no need to change other places. We understand that the numeric number like `1`, `2` and `3` are questions' indexes. My previous review comment mainly focuses on the magic `-1`. It is not so obvious that the `-1` is also a question index.

tshradheya commented:
Sorry, misinterpreted the comment. Updated as requested

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMcqQuestionUiTest.java:
```diff
@@ -59,6 +62,23 @@ public void testNewQuestionFrame() {
+        ______TS("MCQ: Check UI after cancelling and add new MCQ question again");
+
+        feedbackEditPage.clickGenerateMcqOptionsCheckbox(NEW_QUESTION_INDEX);
+        assertFalse(feedbackEditPage.isElementVisible("mcqChoiceTable--1"));
```
How about `mcqChoiceTable- + NEW_QUESTION_INDEX`? Is it more clear? What do u think?

tshradheya replied:
That will require to convert `NEW_QUESTION_INDEX` to string so might get messy. 
Moreover, since its to denote an html tag, I think just writing the whole `id` of element as one string is more clearer as is done in various other places in this file

xpdavid replied:
Alright. Let's leave it there. Note that the string is clear to u because u know that `-1` is the new question index and `-` is the delimiter. I don't think this file doesn't a good job on that.
